homeassistant/components/hassio.py

```diff
@@ -0,0 +1,239 @@
+"""
+Exposes regular rest commands as services.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/hassio/
+"""
+import asyncio
+import logging
+import os
+
+import aiohttp
+import async_timeout
+import voluptuous as vol
+
+from homeassistant.config import load_yaml_config_file
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+
+DOMAIN = 'hassio'
+
+_LOGGER = logging.getLogger(__name__)
+
+TIMEOUT = 900
+DATA_HASSIO = 'hassio'
+
+SERVICE_HOST_SHUTDOWN = 'host_shutdown'
+SERVICE_HOST_REBOOT = 'host_reboot'
+SERVICE_HOST_UPDATE = 'host_update'
+
+SERVICE_NETWORK_OPTIONS = 'network_options'
+
+SERVICE_SUPERVISOR_UPDATE = 'supervisor_update'
+SERVICE_SUPERVISOR_OPTIONS = 'supervisor_options'
+
+SERVICE_HOMEASSISTANT_UPDATE = 'homeassistant_update'
+
+SERVICE_ADDON_INSTALL = 'addon_install'
+SERVICE_ADDON_UNINSTALL = 'addon_uninstall'
+SERVICE_ADDON_UPDATE = 'addon_update'
+SERVICE_ADDON_START = 'addon_start'
+SERVICE_ADDON_STOP = 'addon_stop'
+
+ATTR_ADDON = 'addon'
+ATTR_BETA = 'beta'
+ATTR_OPTIONS = 'options'
+ATTR_IP = 'ip'
+ATTR_SSID = 'ssid'
+ATTR_PASSWORD = 'password'
+ATTR_VERSION = 'version'
+
+
+SCHEMA_SERVICE_UPDATE = vol.Schema({
+    vol.Optional(ATTR_VERSION): cv.string,
+})
+
+SCHEMA_SERVICE_SUPERVISOR_OPTIONS = vol.Schema({
+    vol.Optional(ATTR_BETA): cv.boolean,
+})
+
+SCHEMA_SERVICE_NETWORK_OPTIONS = vol.Schema({
+    vol.Optional(ATTR_IP): cv.string,
+    vol.Optional(ATTR_SSID): cv.string,
+    vol.Optional(ATTR_PASSWORD): cv.string,
+})
+
+SCHEMA_SERVICE_ADDONS = vol.Schema({
+    vol.Required(ATTR_ADDON): cv.slug,
+})
+
+SCHEMA_SERVICE_ADDONS_START = SCHEMA_SERVICE_ADDONS.extend({
+    vol.Optional(ATTR_OPTIONS): dict,
+})
+
+SCHEMA_SERVICE_ADDONS_VERSION = SCHEMA_SERVICE_ADDONS.extend({
+    vol.Optional(ATTR_VERSION): cv.string,
+})
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Setup the hassio component."""
+    websession = async_get_clientsession(hass)
+    hassio = HassIO(hass.loop, websession)
+
+    if not hassio.connected:
+        _LOGGER.error("No HassIO supervisor detect!")
+        return False
+
+    hass.data[DATA_HASSIO] = hassio
+
+    descriptions = yield from hass.loop.run_in_executor(
+        None, load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml'))
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Handle HassIO service calls."""
+        if service.service == SERVICE_HOST_UPDATE:
```

Comment:
I suggest you create a dict that maps services to commands, payloads and schemas and lookup that here. See cover component for example. It will use a lot fewer lines of code.